File: krass_jass/tables.py
"""Precomputed lookup tables.

The engine does no rule *reasoning* in its hot paths — it indexes into these. Everything
here is derived once at import from :mod:`krass_jass.cards` and the value/order tables in
``docs/rules-config.md``, and everything is a plain tuple of ints so the rollout kernel can
stay free of Python objects (see ``docs/plan-review.md`` §1 on why that matters).
"""

from __future__ import annotations

from .cards import (
    NUM_CARDS,
    NUM_RANKS,
    NUM_SUITS,
    RANK_6,
    RANK_7,
    RANK_8,
    RANK_9,
    RANK_A,
    RANK_J,
    RANK_K,
    RANK_Q,
    RANK_T,
    card_rank,
    card_suit,
)
from .rules import Contract

# --- card values, per rank -------------------------------------------------

_VAL_PLAIN = (11, 4, 3, 2, 10, 0, 0, 0, 0)          # A K Q J 10 9 8 7 6
_VAL_TRUMP = (11, 4, 3, 20, 10, 14, 0, 0, 0)        # Puur 20, Näll 14
_VAL_OBENABE = (11, 4, 3, 2, 10, 0, 8, 0, 0)        # 8s pay 8
_VAL_UNDENUFE = (0, 4, 3, 2, 10, 0, 8, 0, 11)       # values invert with the order: 6=11, A=0

# --- strength, per rank (higher wins) --------------------------------------

_STR_PLAIN = tuple(NUM_RANKS - 1 - r for r in range(NUM_RANKS))        # A high
_STR_UNDENUFE = tuple(range(NUM_RANKS))                                # 6 high
#: J > 9 > A > K > Q > 10 > 8 > 7 > 6
_STR_TRUMP = tuple(
    {RANK_J: 8, RANK_9: 7, RANK_A: 6, RANK_K: 5, RANK_Q: 4, RANK_T: 3, RANK_8: 2, RANK_7: 1, RANK_6: 0}[r]
    for r in range(NUM_RANKS)
)

#: ``TRUMP_HIGHER[s]`` — 9-bit mask of trump *ranks* strictly stronger than trump strength
#: ``s``. Shift by ``trump * 9`` to get a card mask. This is what makes the undertrump rule
#: a lookup rather than a comparison loop.
TRUMP_HIGHER: tuple[int, ...] = tuple(
    sum(1 << r for r in range(NUM_RANKS) if _STR_TRUMP[r] > s) for s in range(NUM_RANKS)
)

#: Mask of the Puur, per trump suit.
PUUR_MASK: tuple[int, ...] = tuple(1 << (s * NUM_RANKS + RANK_J) for s in range(NUM_SUITS))

#: Mask of King+Queen of trumps (Stöck), per trump suit.
STOECK_MASK: tuple[int, ...] = tuple(
    (1 << (s * NUM_RANKS + RANK_K)) | (1 << (s * NUM_RANKS + RANK_Q)) for s in range(NUM_SUITS)
)


def _values_for(contract: Contract) -> tuple[int, ...]:
    """Point value of every card index, for one contract."""
    if contract is Contract.OBENABE:
        per_rank = [_VAL_OBENABE] * NUM_SUITS
    elif contract is Contract.UNDENUFE:
        per_rank = [_VAL_UNDENUFE] * NUM_SUITS
    else:
        trump = contract.trump_suit
        per_rank = [_VAL_TRUMP if s == trump else _VAL_PLAIN for s in range(NUM_SUITS)]
    return tuple(per_rank[card_suit(c)][card_rank(c)] for c in range(NUM_CARDS))


def _strength_for(contract: Contract, led: int) -> tuple[int, ...]:
    """Trick strength of every card index, given the led suit.

    Banded so a single ``max`` decides the trick: trump 100+, led suit 10+, anything else
    0. A card that is neither trump nor of the led suit can never win, and the leader's own
    card is always in a scoring band, so the max is never ambiguous.
    """
    trump = contract.trump_suit
    base = _STR_UNDENUFE if contract is Contract.UNDENUFE else _STR_PLAIN
    out = []
    for c in range(NUM_CARDS):
        s, r = card_suit(c), card_rank(c)
        if s == trump:
            out.append(100 + _STR_TRUMP[r])
        elif s == led:
            out.append(10 + base[r])
        else:
            out.append(0)
    return tuple(out)


#: ``CARD_VALUES[contract]`` -> 36 point values.
CARD_VALUES: dict[Contract, tuple[int, ...]] = {c: _values_for(c) for c in Contract}

#: ``STRENGTH[contract][led_suit]`` -> 36 strengths.
STRENGTH: dict[Contract, tuple[tuple[int, ...], ...]] = {
    c: tuple(_strength_for(c, led) for led in range(NUM_SUITS)) for c in Contract
}

#: Points in the whole deck, per contract. Every contract totals 152 — which is exactly why
#: a "sums to 157" test cannot catch a mis-implemented Undenufe.
DECK_POINTS: dict[Contract, int] = {c: sum(v) for c, v in CARD_VALUES.items()}

File: rust/src/tables.rs
//! Precomputed value and strength tables, mirroring `krass_jass/tables.py`.
//!
//! Everything is `const` — built at compile time with while-loops — so the hot paths index
//! into static arrays and the search never recomputes a rule.

use crate::cards::*;

pub const NUM_CONTRACTS: usize = 6;
pub const OBENABE: usize = 4;
pub const UNDENUFE: usize = 5;

const VAL_PLAIN: [i32; NUM_RANKS] = [11, 4, 3, 2, 10, 0, 0, 0, 0];
const VAL_TRUMP: [i32; NUM_RANKS] = [11, 4, 3, 20, 10, 14, 0, 0, 0];
const VAL_OBENABE: [i32; NUM_RANKS] = [11, 4, 3, 2, 10, 0, 8, 0, 0];
/// Undenufe inverts the values along with the order: the 6 is worth 11, the ace 0.
const VAL_UNDENUFE: [i32; NUM_RANKS] = [0, 4, 3, 2, 10, 0, 8, 0, 11];

/// J > 9 > A > K > Q > 10 > 8 > 7 > 6
const STR_TRUMP: [i32; NUM_RANKS] = [6, 5, 4, 8, 3, 7, 2, 1, 0];

const fn str_plain(r: usize) -> i32 {
    (NUM_RANKS - 1 - r) as i32
}

/// `TRUMP_HIGHER[s]` — mask of trump *ranks* strictly stronger than trump strength `s`.
/// Shift by `trump * 9` for a card mask. Makes the undertrump rule a lookup.
const fn build_trump_higher() -> [u64; NUM_RANKS] {
    let mut out = [0u64; NUM_RANKS];
    let mut s = 0;
    while s < NUM_RANKS {
        let mut r = 0;
        while r < NUM_RANKS {
            if STR_TRUMP[r] > s as i32 {
                out[s] |= 1u64 << r;
            }
            r += 1;
        }
        s += 1;
    }
    out
}

pub const TRUMP_HIGHER: [u64; NUM_RANKS] = build_trump_higher();

const fn trump_of(contract: usize) -> i32 {
    if contract < 4 {
        contract as i32
    } else {
        -1
    }
}

const fn build_values() -> [[i32; NUM_CARDS]; NUM_CONTRACTS] {
    let mut out = [[0i32; NUM_CARDS]; NUM_CONTRACTS];
    let mut k = 0;
    while k < NUM_CONTRACTS {
        let trump = trump_of(k);
        let mut c = 0;
        while c < NUM_CARDS {
            let s = c / NUM_RANKS;
            let r = c % NUM_RANKS;
            out[k][c] = if k == OBENABE {
                VAL_OBENABE[r]
            } else if k == UNDENUFE {
                VAL_UNDENUFE[r]
            } else if s as i32 == trump {
                VAL_TRUMP[r]
            } else {
                VAL_PLAIN[r]
            };
            c += 1;
        }
        k += 1;
    }
    out
}

pub const CARD_VALUES: [[i32; NUM_CARDS]; NUM_CONTRACTS] = build_values();

/// `STRENGTH[contract][led][card]`, banded so one `max` decides the trick:
/// trump 100+, led suit 10+, anything else 0.
const fn build_strength() -> [[[i32; NUM_CARDS]; NUM_SUITS]; NUM_CONTRACTS] {
    let mut out = [[[0i32; NUM_CARDS]; NUM_SUITS]; NUM_CONTRACTS];
    let mut k = 0;
    while k < NUM_CONTRACTS {
        let trump = trump_of(k);
        let mut led = 0;
        while led < NUM_SUITS {
            let mut c = 0;
            while c < NUM_CARDS {
                let s = c / NUM_RANKS;
                let r = c % NUM_RANKS;
                out[k][led][c] = if s as i32 == trump {
                    100 + STR_TRUMP[r]
                } else if s == led {
                    // Undenufe reverses the order: the 6 is high
                    10 + if k == UNDENUFE { r as i32 } else { str_plain(r) }
                } else {
                    0
                };
                c += 1;
            }
            led += 1;
        }
        k += 1;
    }
    out
}

pub const STRENGTH: [[[i32; NUM_CARDS]; NUM_SUITS]; NUM_CONTRACTS] = build_strength();

/// `SUIT_ORDER[contract][suit]` — the nine card indices of that suit, strongest first.
///
/// Used by the endgame solver's equivalence reduction, which needs the ranking *within* a
/// suit independent of what was led.
const fn build_suit_order() -> [[[usize; NUM_RANKS]; NUM_SUITS]; NUM_CONTRACTS] {
    let mut out = [[[0usize; NUM_RANKS]; NUM_SUITS]; NUM_CONTRACTS];
    let mut k = 0;
    while k < NUM_CONTRACTS {
        let trump = trump_of(k);
        let mut suit = 0;
        while suit < NUM_SUITS {
            // insertion sort by strength, descending
            let mut n = 0;
            while n < NUM_RANKS {
                let mut best_r = usize::MAX;
                let mut best_s = -1i32;
                let mut r = 0;
                while r < NUM_RANKS {
                    // skip ranks already placed
                    let mut placed = false;
                    let mut i = 0;
                    while i < n {
                        if out[k][suit][i] == suit * NUM_RANKS + r {
                            placed = true;
                        }
                        i += 1;
                    }
                    if !placed {
                        let st = if suit as i32 == trump {
                            STR_TRUMP[r]
                        } else if k == UNDENUFE {
                            r as i32
                        } else {
                            str_plain(r)
                        };
                        if st > best_s {
                            best_s = st;
                            best_r = r;
                        }
                    }
                    r += 1;
                }
                out[k][suit][n] = suit * NUM_RANKS + best_r;
                n += 1;
            }
            suit += 1;
        }
        k += 1;
    }
    out
}

pub const SUIT_ORDER: [[[usize; NUM_RANKS]; NUM_SUITS]; NUM_CONTRACTS] = build_suit_order();
